### PR TITLE
Ke git 6512 forms features

### PR DESCRIPTION
**Pull request by keichwa**
https://github.com/SUSE/spacewalk/issues/6512
and @mvidner, please review.

### CODE REVIEW

File: adoc/bp_chap_getting_started_with_salt_formulas.adoc [hunk 1]
```diff
@@ -331,13 +331,43 @@ Many of these values are self-explanatory:
...
 The difference between `group` and `namespace` is `group` generates a visible border with a heading, and `namespace` shows nothing visually (and is only used to structure pillar data).
 The difference between `group` and `edit-group` is: `edit-group` allows to structure and restrict editable fields in a more flexible way.
+`edit-group` is a collection of items of the same kind; collections can have the following four "shapes":
+
+** A list of primitive items
+** A list of dictionaries
+** A dictionary of primitive items
+** A dictionary of dictionaries
+
+The size of such collection is variable; users can add or remove elements.
```
Comment (Loquacity) [inline adoc/bp_chap_getting_started_with_salt_formulas.adoc]: s/such/each ?

File: adoc/bp_chap_getting_started_with_salt_formulas.adoc [hunk 2]
```diff
@@ -331,13 +331,43 @@ Many of these values are self-explanatory:
...
 For an `edit-group` example, see <<salt.formulas.ex.edit-group>>.
 
 $default::
 `$default` allows you to specify a default value that is displayed and used, if no other value is entered.
 In an `edit-group` it allows to create initial members of the group and populate them with specified data.
 
+$optional::
+`$optional` is a boolean attribute.
+If it is `true` and the field is empty in the form, then this field will not be generated in the formula data.
+The generated dictionary will miss the field name key.
```
Comment (Loquacity) [inline adoc/bp_chap_getting_started_with_salt_formulas.adoc]: Is there a more specific word than 'miss'? Possibly 'skip'?

File: adoc/bp_chap_getting_started_with_salt_formulas.adoc [hunk 3]
```diff
@@ -351,14 +381,45 @@ The `$help` and `$placeholder` attributes are used to give a user a better under
 `$placeholder` may only be used with text fields like text, password, email or date.
 It does not make sense to add a placeholder if you also use `$default` as this will hide the placeholder.
 
+$key::
+`$key` is applicable if the `edit-group` has a "shape" of a dictionary; you use it when the pillar data is supposed to be a dictionary.
```
Comment (Loquacity) [inline adoc/bp_chap_getting_started_with_salt_formulas.adoc]: s/a "shape"/the "shape"

File: adoc/bp_chap_getting_started_with_salt_formulas.adoc [hunk 4]
```diff
@@ -351,14 +381,45 @@ The `$help` and `$placeholder` attributes are used to give a user a better under
...
+
 $minItems and $maxItems::
 In an ``edit-group``, `$minItems` and `$maxItems` allow you to specify the lowest and highest number the group can occur.
 
 $itemName::
 In an ``edit-group``, `$itemName` allows to define a template for the name to be used for the members of the group.
 
 $prototype::
-In an ``edit-group``, `$prototype` allows to define default (or pre-filled) values for newly added members in the group.
+In an ``edit-group``, `$prototype` is mandatory and allows to define default (or pre-filled) values for newly added members in the group.
```
Comment (Loquacity) [inline adoc/bp_chap_getting_started_with_salt_formulas.adoc]: "and allows you to"

File: adoc/bp_chap_getting_started_with_salt_formulas.adoc (hunk 2 above)
Comment (keichwa) [reply adoc/bp_chap_getting_started_with_salt_formulas.adoc]: I tried a different version. @hustodemon can you please check whether it makes sense?

File: adoc/bp_chap_getting_started_with_salt_formulas.adoc [hunk 5]
```diff
@@ -331,13 +331,42 @@ Many of these values are self-explanatory:
...
 For an `edit-group` example, see <<salt.formulas.ex.edit-group>>.
 
 $default::
 `$default` allows you to specify a default value that is displayed and used, if no other value is entered.
 In an `edit-group` it allows to create initial members of the group and populate them with specified data.
 
+$optional::
+`$optional` is a boolean attribute.
+If it is `true` and the field is empty in the form, then this field will not be generated in the formula data and the generated dictionary will not contain the field name key.
+If `$optional` is `false`, the formula data will contain a `<field name>: null` entry.
```
Comment (hustodemon) [inline adoc/bp_chap_getting_started_with_salt_formulas.adoc]: Maybe I'd explicitly mention "if `$optional` is false and the field is empty. Feel free to ignore if you think this is clear enough already.

**Comment by keichwa** [comment]
Frantisek Kobzik <notifications@github.com> writes:

> Maybe I'd explicitly mention "if `$optional` is false and the field is
> empty. Feel free to ignore if you think this is clear enough already.

Thanks for reviewing!

For clarity, it makes sense.  I hope I got it right.

-- 
Karl Eichwalder            SUSE Linux GmbH
PES / Documentation        Maxfeldstraße 5
                           90409 Nürnberg, Germany
GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg)

**Pull request merged**